### PR TITLE
Dynamic Virtual workspace implementation

### PR DESCRIPTION
**Pull request by davidfestal**
Dynamic Virtual workspace implementation that allows exposing CRD-like APIs per API domains
(one base virtual workspace URL per API domain)

Largly inspired from the current code of the `apiextensions-apiserver/pkg/apiserver`
package of the `kcp-dev/kubernetes` fork

This is a pre-requisite for https://github.com/kcp-dev/kcp/pull/754.

### CODE REVIEW

File: pkg/virtual/framework/dynamic/doc_test.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	genericapiserver "k8s.io/apiserver/pkg/server"
+
+	"github.com/kcp-dev/kcp/pkg/apis/apiresource/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic"
+	apidefs "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/apidefs"
+	"github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/apiserver"
+)
+
+// Typical Usage of a DynamicVirtualWorkspace
+func Example() {
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/doc_test.go:32]: doc_test.go looks odd. pkg/virtual/framework/dynamic/example or so sounds better.

File: pkg/virtual/framework/dynamic/doc_test.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+			if _, exists := someAPISetRetriever.GetAPIs(apiDomainKey); !exists {
+				// If the APISetRetriever doesn't manage this api domain, don't accept the request
+				return
+			}
+
+			// ...
+
+			// Add the apiDomainKey to the request context before passing the request to the virtual workspace APIServer
+			completedContext = context.WithValue(requestContext, apidefs.APIDomainKeyContextKey, apiDomainKey)
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/doc_test.go:59]: can we expose `WithAPIDomainKey` and `APIDomainKeyFrom` instead of the actual key? The former is what we have everywhere else including kube.

File: pkg/virtual/framework/dynamic/doc_test.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+				// If the APISetRetriever is not initialized, don't accept the request
+				return
+			}
+
+			var apiDomainKey string
+
+			// Resolve the request root path, and extract API domain key from it ...
+
+			if _, exists := someAPISetRetriever.GetAPIs(apiDomainKey); !exists {
+				// If the APISetRetriever doesn't manage this api domain, don't accept the request
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/doc_test.go:52]: that's an odd place here in the root resolver.

File: pkg/virtual/framework/dynamic/apidefs/types.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	apiresourcev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apiresource/v1alpha1"
+)
+
+// APIDomainKeyContextKeyType is the type of the key for the request context value
+// that will carry the API domain key.
+type APIDomainKeyContextKeyType string
+
+// APIDomainKeyContextKey is the key for the request context value
+// that will carry the API domain key.
+const APIDomainKeyContextKey APIDomainKeyContextKeyType = "VirtualWorkspaceAPIDomainLocationKey"
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/apidefs/types.go:33]: APIDomainLocation is hard to understand. What is a location, what a domain?

Also, please expose `WithFoo(ctx, ...)` and `FooFrom(ctx)`  instead of the keys.

Why is this here in apidefs? Why does a context have something an APIDefinition or vice-versa?

File: pkg/virtual/framework/dynamic/apidefs/types.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+// APIDefinition provides access to all the information needed to serve a given API resource
+type APIDefinition interface {
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/apidefs/types.go:36]: What is an API? Are you talking about a resource? ResourceDefinition is a better fit?

File: pkg/virtual/framework/dynamic/apidefs/types.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	// GetAPIResourceSpec provides the API resource specification, which contains the
+	// API names, sub-resource definitions, and the OpenAPIv3 schema.
+	GetAPIResourceSpec() *apiresourcev1alpha1.CommonAPIResourceSpec
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/apidefs/types.go:39]: can this only be backed by APIResources?

File: pkg/virtual/framework/dynamic/apidefs/types.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	GetClusterName() string
+
+	// GetStorage provides the REST storage used to serve the resource.
+	GetStorage() rest.Storage
+
+	// GetSubResourceStorage provides the REST storage required to serve the given sub-resource.
+	GetSubResourceStorage(subresource string) rest.Storage
+
+	// GetRequestScope provides the handlers.RequestScope required to serve the resource.
+	GetRequestScope() *handlers.RequestScope
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/apidefs/types.go:51]: why a pointer?

File: pkg/virtual/framework/dynamic/apidefs/types.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+	// GetSubResourceRequestScope provides the handlers.RequestScope required to serve the given sub-resource.
+	GetSubResourceRequestScope(subresource string) *handlers.RequestScope
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/apidefs/types.go:54]: How can this be different from the main resource?

File: pkg/virtual/framework/dynamic/apidefs/types.go (hunk 7 above)
Comment (sttts) [reply pkg/virtual/framework/dynamic/apidefs/types.go:51]: nvmd. This is a bad name for that type (in upstream).

File: pkg/virtual/framework/dynamic/apidefs/types.go (hunk 8 above)
Comment (sttts) [reply pkg/virtual/framework/dynamic/apidefs/types.go:54]: same

File: pkg/virtual/framework/dynamic/apidefs/types.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+// APISet contains the APIDefintion objects for the APIs of an API domain.
+type APISet map[schema.GroupVersionResource]APIDefinition
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/apidefs/types.go:58]: ResourceDefinitionSet

File: pkg/virtual/framework/dynamic/apiserver/apiserver.go [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+	"k8s.io/apiserver/pkg/endpoints/discovery"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+
+	virtualcontext "github.com/kcp-dev/kcp/pkg/virtual/framework/context"
+	apidefs "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/apidefs"
+)
+
+var (
+	Scheme = runtime.NewScheme()
+	Codecs = serializer.NewCodecFactory(Scheme)
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/apiserver/apiserver.go:38]: commented before: no public schemes please.

File: pkg/virtual/framework/dynamic/apiserver/apiserver.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
+/*
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/apiserver/apiserver.go:1]: apiserver.go is a bad file name. This is usally called `config.go`.

File: pkg/virtual/framework/dynamic/apiserver/doc.go [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+//
+// The main interface between this package and other components is the apidefs.APIDefinition:
+//
+// - the configuration of the DynamicAPIServer contains an apidefs.APISetRetriever that allows retrieving an apidefs.APIDefinition
+// based on an api domain key and a GVR
+//
+// - the CreateServingInfoFor method can be used by external components at any time to create an apidefs.APIDefinition and
+// add it to the apidefs.APISetRetriever that has been passed to the DynamicAPIServer
+//
+// Parts of this package are highly insired from kcp-dev/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/apiserver/doc.go:28]:
```suggestion
// Parts of this package are highly insired from k8s.io/apiextensions-apiserver/pkg/apiserver
```

File: pkg/virtual/framework/dynamic/apiserver/handler.go [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+		Resource: requestInfo.Resource,
+	}]
+
+	if !exists {
+		r.delegate.ServeHTTP(w, req)
+		return
+	}
+
+	apiResourceSpec := apiDef.GetAPIResourceSpec()
+	//	apiClusterName := apiDef.GetClusterName()
```
Comment (sttts) [inline pkg/virtual/framework/dynamic/apiserver/handler.go:146]: remove?

File: pkg/virtual/framework/dynamic/apidefs/types.go (hunk 6 above)
Comment (sttts) [reply pkg/virtual/framework/dynamic/apidefs/types.go:39]: looked further down at the forked handler code it looks like returning a CRD here would make the handler code much more similar to apiextensions-apiserver, potentially even allow to reuse the handler.

Comment (sttts) [reply pkg/virtual/framework/dynamic/apidefs/types.go:39]: cc @ncdc

Comment (davidfestal) [reply pkg/virtual/framework/dynamic/apidefs/types.go:39]: The main point is that we don't need all the multi-version aspects here.

Comment (davidfestal) [reply pkg/virtual/framework/dynamic/apidefs/types.go:39]: But I assume we could also wrap this info into a CRD if we want to reuse code. We've been discussing the 2 options a while ago.

File: pkg/virtual/framework/dynamic/apiserver/apiserver.go (hunk 10 above)
Comment (davidfestal) [reply pkg/virtual/framework/dynamic/apiserver/apiserver.go:38]: Oh, wait, I didn't add the related fixes of the initial PR. Let me add those.

File: pkg/virtual/framework/dynamic/doc_test.go (hunk 1 above)
Comment (davidfestal) [reply pkg/virtual/framework/dynamic/doc_test.go:34]: I can name it `example_test.go`, which might make it a bit clearer if you want.

File: pkg/virtual/framework/dynamic/doc_test.go (hunk 3 above)
Comment (davidfestal) [reply pkg/virtual/framework/dynamic/doc_test.go:54]: Why ? wdym ?

File: pkg/virtual/framework/dynamic/apiserver/apiserver.go (hunk 10 above)
Comment (davidfestal) [reply pkg/virtual/framework/dynamic/apiserver/apiserver.go]: Fixed by reintegrating the prior fixes of PR #754

File: pkg/virtual/framework/dynamic/apidefs/types.go (hunk 5 above)
Comment (davidfestal) [reply pkg/virtual/framework/dynamic/apidefs/types.go]: Still not convinced here, especially if we use a raw CRD instead of `CommonAPIResourceSpec`. I'd rather leave it for when  the refactoring mentioned here is done / validated.

File: pkg/virtual/framework/dynamic/apidefs/types.go (hunk 9 above)
Comment (davidfestal) [reply pkg/virtual/framework/dynamic/apidefs/types.go]: see https://github.com/kcp-dev/kcp/pull/1039#discussion_r871535583

File: pkg/virtual/framework/dynamic/doc_test.go (hunk 2 above)
Comment (davidfestal) [reply pkg/virtual/framework/dynamic/doc_test.go]: Fixed in commit 23a45c6ae943eb04780a29a8b10d28ac10b24b2d

File: pkg/virtual/framework/dynamic/apiserver/doc.go (hunk 12 above)
Comment (davidfestal) [reply pkg/virtual/framework/dynamic/apiserver/doc.go]: Fixed in commit 23a45c6ae943eb04780a29a8b10d28ac10b24b2d

**Review by sttts — APPROVED** [review]

File: pkg/virtual/framework/dynamic/apidefs/types.go (hunk 5 above)
Comment (sttts) [reply pkg/virtual/framework/dynamic/apidefs/types.go]: Am convinced that Resource is a better term :-) But as discussed we can wait until the APIBinding refactor.

File: pkg/virtual/framework/dynamic/apidefs/types.go (hunk 6 above)
Comment (sttts) [reply pkg/virtual/framework/dynamic/apidefs/types.go]: We discussed this and will postpone until the APIBinding refactor.
